### PR TITLE
Add ReflectVisitor and ResolvedReferences

### PR DESCRIPTION
**Pull request by matthchr**
Laying the groundwork for resolving arbitrary cross resource references.

### CODE REVIEW

File: hack/generated/pkg/reflecthelpers/reflect_visitor.go [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+
+type ReflectVisitor struct {
+	VisitPrimitive func(this *ReflectVisitor, it interface{}, ctx interface{}) error
+	VisitStruct    func(this *ReflectVisitor, it interface{}, ctx interface{}) error
+	VisitPtr       func(this *ReflectVisitor, it interface{}, ctx interface{}) error
+	VisitSlice     func(this *ReflectVisitor, it interface{}, ctx interface{}) error
+	VisitMap       func(this *ReflectVisitor, it interface{}, ctx interface{}) error
+}
+
+func NewReflectVisitor() *ReflectVisitor {
```
Comment (matthchr) [inline hack/generated/pkg/reflecthelpers/reflect_visitor.go:52]: Missing comments for many of these functions

File: hack/generated/pkg/genruntime/resolved_references.go [hunk 2]
```diff
@@ -0,0 +1,29 @@
+/*
+ * Copyright (c) Microsoft Corporation.
+ * Licensed under the MIT license.
+ */
+
+package genruntime
+
+type ResolvedReferences struct {
```
Comment (matthchr) [inline hack/generated/pkg/genruntime/resolved_references.go:8]: docs

File: hack/generated/pkg/genruntime/resolved_references.go [hunk 3]
```diff
@@ -0,0 +1,39 @@
+/*
+ * Copyright (c) Microsoft Corporation.
+ * Licensed under the MIT license.
+ */
+
+package genruntime
+
+// ResolvedReferences is a set of references which have been resolved for a particular resource.
+// The special self field is the fully qualified ARM ID of the resource that this ResolvedReferences applies to.
+type ResolvedReferences struct {
```
Comment (babbageclunk) [inline hack/generated/pkg/genruntime/resolved_references.go:10]: I don't get why you're bundling these two things together into one struct. Why is name in here?

**Review by babbageclunk — APPROVED** [review]
I like this, but I think name should be a separate argument to ConvertToARM, rather than packaging it up with the resolved references.

The reflect-based visitor is nice! I've seen code in other projects that uses reflection to gather info from arbitrary types in similar but in a much more ad-hoc way - this is much better.

File: hack/generated/pkg/genruntime/resolved_references.go (hunk 3 above)
Comment (matthchr) [reply hack/generated/pkg/genruntime/resolved_references.go:10]: Pulled it back out

Comment (matthchr) [reply hack/generated/pkg/genruntime/resolved_references.go:10]: Removed name from this

**Pull request merged**